pkg/cmd/roachtest/tests/lease.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func registerLease(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:    "acquire-lease/crash/non-system",
+		Owner:   registry.OwnerKV,
+		Timeout: time.Hour,
+		Cluster: r.MakeClusterSpec(7, spec.CPU(4)),
+		Run:     runAcquireLeaseCrashNonSystem,
+	})
+}
+
+// runAcquireLeaseCrashNonSystem benchmarks the maximum duration of range
+// unavailability following a leaseholder crash. It tests the simplest possible
+// failure:
+//
+//   - A process crash, where the host/OS remains available (in particular, the
+//     TCP/IP stack is responsive and sends immediate RST packets to peers).
+//
+//   - No system ranges located on the crashed node.
+//
+//   - SQL clients do not connect to the crashed node.
+//
+//   - The workload consists of individual point reads and writes.
+//
+// We expect the worst-case (pMax) latency to be slightly larger than the lease
+// interval (9s), to account for lease acquisition and retry latencies. We do
+// not assert this, but instead export latency histograms for graphing.
+//
+// The cluster layout is as follows:
+//
+// n1-n3: System ranges and SQL gateways.
+// n4-n6: Workload ranges.
+// n7:    Workload runner.
+//
+// The test runs a kv50 workload with batch size 1, using 256 concurrent
+// workers directed at n1-n3. n4-n6 are killed and restarted in order, with 30
+// seconds between each operation, for 3 cycles totaling 9 crashes.
+func runAcquireLeaseCrashNonSystem(ctx context.Context, t test.Test, c cluster.Cluster) {
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	c.Put(ctx, t.DeprecatedWorkload(), "./workload")
+
+	// Create cluster with one rack locality per node and configure it.
+	require.Equal(t, 7, c.Spec().NodeCount)
+
+	opts := option.DefaultStartOpts()
+	settings := install.MakeClusterSettings(install.NumRacksOption(c.Spec().NodeCount))
+	c.Start(ctx, t.L(), opts, settings, c.Range(1, 6))
+
+	conn := c.Conn(ctx, t.L(), 1)
+
+	t.Status("configuring cluster")
+	_, err := conn.ExecContext(ctx, `SET CLUSTER SETTING kv.range_split.by_load_enabled = 'false'`)
+	require.NoError(t, err)
+
+	// Constrain all existing zone configs to n1-n3.
+	conn2 := c.Conn(ctx, t.L(), 1)
+	rows, err := conn.QueryContext(ctx, `SELECT target FROM [SHOW ALL ZONE CONFIGURATIONS]`)
+	require.NoError(t, err)
+	for rows.Next() {
+		var target string
+		require.NoError(t, rows.Scan(&target))
+		_, err = conn2.ExecContext(ctx, fmt.Sprintf(
+			`ALTER %s CONFIGURE ZONE USING num_replicas = 3, constraints = '[-rack=3, -rack=4, -rack=5]'`,
+			target))
+		require.NoError(t, err)
+	}
+	require.NoError(t, rows.Err())
+	rows.Close()
+	require.NoError(t, conn2.Close())
+
+	// Wait for upreplication.
```

Comment:
Note that this doesn't wait for the 5x replication expected for system ranges here.

Reply:
We explicitly set them to 3x replication above. We don't want them on the workload nodes.